File: node/src/security.mjs
import {
  argon2Sync,
  createHash,
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  randomBytes,
  timingSafeEqual,
} from 'node:crypto';

const ARGON2_MEMORY_KIB = 65_536;
const ARGON2_PASSES = 3;
const ARGON2_PARALLELISM = 1;
const ARGON2_TAG_LENGTH = 32;

export async function hashPassword(password) {
  validatePassword(password);
  const salt = randomBytes(16);
  const digest = argon2Sync('argon2id', {
    message: Buffer.from(password, 'utf8'),
    nonce: salt,
    parallelism: ARGON2_PARALLELISM,
    tagLength: ARGON2_TAG_LENGTH,
    memory: ARGON2_MEMORY_KIB,
    passes: ARGON2_PASSES,
  });
  return [
    'argon2id',
    'v=19',
    `m=${ARGON2_MEMORY_KIB},t=${ARGON2_PASSES},p=${ARGON2_PARALLELISM}`,
    salt.toString('base64url'),
    Buffer.from(digest).toString('base64url'),
  ].join('$');
}

export async function verifyPassword(password, encoded) {
  try {
    const [algorithm, version, parameters, saltText, digestText] = encoded.split('$');
    if (algorithm !== 'argon2id' || version !== 'v=19') return false;
    const parsed = Object.fromEntries(
      parameters.split(',').map((item) => {
        const [key, value] = item.split('=');
        return [key, Number(value)];
      }),
    );
    const expected = Buffer.from(digestText, 'base64url');
    const salt = Buffer.from(saltText, 'base64url');
    if (
      parsed.m !== ARGON2_MEMORY_KIB
      || parsed.t !== ARGON2_PASSES
      || parsed.p !== ARGON2_PARALLELISM
      || expected.length !== ARGON2_TAG_LENGTH
      || salt.length !== 16
    ) return false;
    const actual = Buffer.from(
      argon2Sync('argon2id', {
        message: Buffer.from(password, 'utf8'),
        nonce: salt,
        parallelism: parsed.p,
        tagLength: expected.length,
        memory: parsed.m,
        passes: parsed.t,
      }),
    );
    return expected.length === actual.length && timingSafeEqual(expected, actual);
  } catch {
    return false;
  }
}

export function validatePassword(password) {
  if (typeof password !== 'string' || password.length < 12 || password.length > 256) {
    throw new Error('Password must contain between 12 and 256 characters');
  }
}

export function generateToken() {
  return randomBytes(32).toString('base64url');
}

export function tokenHash(token) {
  return createHash('sha256').update(token, 'utf8').digest('hex');
}

export function generateNodeIdentity() {
  const { privateKey, publicKey } = generateKeyPairSync('ed25519');
  const publicKeyDer = publicKey.export({ type: 'spki', format: 'der' });
  return {
    nodeId: `srcnode_${createHash('sha256').update(publicKeyDer).digest('base64url')}`,
    publicKey: publicKeyDer.toString('base64url'),
    privateKey: privateKey.export({ type: 'pkcs8', format: 'pem' }).toString(),
  };
}

export function parseEd25519PublicKey(encoded) {
  if (typeof encoded !== 'string' || encoded.length < 40 || encoded.length > 256) {
    throw new Error('Invalid Ed25519 public key');
  }
  const der = Buffer.from(encoded, 'base64url');
  const key = createPublicKey({ key: der, type: 'spki', format: 'der' });
  if (key.asymmetricKeyType !== 'ed25519') throw new Error('Invalid Ed25519 public key');
  const canonical = key.export({ type: 'spki', format: 'der' });
  if (!canonical.equals(der)) throw new Error('Invalid Ed25519 public key');
  return { key, der, encoded: canonical.toString('base64url') };
}

export function loadNodePrivateKey(encoded) {
  const key = createPrivateKey(encoded);
  if (key.asymmetricKeyType !== 'ed25519') throw new Error('Invalid Node private key');
  return key;
}

export function clientIdFromPublicKey(publicKeyDer) {
  return `srcclient_${createHash('sha256').update(publicKeyDer).digest('base64url')}`;
}

export function safeTokenHashEqual(token, expectedHash) {
  if (typeof token !== 'string' || typeof expectedHash !== 'string' || !/^[0-9a-f]{64}$/.test(expectedHash)) return false;
  const actual = Buffer.from(tokenHash(token), 'hex');
  const expected = Buffer.from(expectedHash, 'hex');
  return timingSafeEqual(actual, expected);
}

File: node/internal/security/security.go
package security

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"encoding/pem"
	"errors"
	"fmt"
	"strings"

	"golang.org/x/crypto/argon2"
)

const (
	argonMemory      = 65_536
	argonPasses      = 3
	argonParallelism = 1
	argonTagLength   = 32
)

var rawURL = base64.RawURLEncoding

type NodeIdentity struct{ NodeID, PublicKey, PrivateKey string }

func ValidatePassword(password string) error {
	if len(password) < 12 || len(password) > 256 {
		return errors.New("Password must contain between 12 and 256 characters")
	}
	return nil
}

func HashPassword(password string) (string, error) {
	if err := ValidatePassword(password); err != nil {
		return "", err
	}
	salt := make([]byte, 16)
	if _, err := rand.Read(salt); err != nil {
		return "", err
	}
	digest := argon2.IDKey([]byte(password), salt, argonPasses, argonMemory, argonParallelism, argonTagLength)
	return fmt.Sprintf("argon2id$v=19$m=%d,t=%d,p=%d$%s$%s", argonMemory, argonPasses, argonParallelism, rawURL.EncodeToString(salt), rawURL.EncodeToString(digest)), nil
}

func VerifyPassword(password, encoded string) bool {
	parts := strings.Split(encoded, "$")
	if len(parts) != 5 || parts[0] != "argon2id" || parts[1] != "v=19" {
		return false
	}
	var memory uint32
	var passes uint32
	var parallelism uint8
	if _, err := fmt.Sscanf(parts[2], "m=%d,t=%d,p=%d", &memory, &passes, &parallelism); err != nil {
		return false
	}
	if memory != argonMemory || passes != argonPasses || parallelism != argonParallelism {
		return false
	}
	salt, err := rawURL.DecodeString(parts[3])
	if err != nil || len(salt) != 16 {
		return false
	}
	expected, err := rawURL.DecodeString(parts[4])
	if err != nil || len(expected) != argonTagLength {
		return false
	}
	actual := argon2.IDKey([]byte(password), salt, passes, memory, parallelism, uint32(len(expected)))
	return subtle.ConstantTimeCompare(actual, expected) == 1
}

func GenerateToken() (string, error) {
	value := make([]byte, 32)
	if _, err := rand.Read(value); err != nil {
		return "", err
	}
	return rawURL.EncodeToString(value), nil
}
func TokenHash(token string) string {
	sum := sha256.Sum256([]byte(token))
	return hex.EncodeToString(sum[:])
}

func SafeTokenHashEqual(token, expectedHash string) bool {
	if len(expectedHash) != 64 {
		return false
	}
	expected, err := hex.DecodeString(expectedHash)
	if err != nil {
		return false
	}
	actual := sha256.Sum256([]byte(token))
	return subtle.ConstantTimeCompare(actual[:], expected) == 1
}

func GenerateNodeIdentity() (NodeIdentity, error) {
	publicKey, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return NodeIdentity{}, err
	}
	publicDER, err := x509.MarshalPKIXPublicKey(publicKey)
	if err != nil {
		return NodeIdentity{}, err
	}
	privateDER, err := x509.MarshalPKCS8PrivateKey(privateKey)
	if err != nil {
		return NodeIdentity{}, err
	}
	sum := sha256.Sum256(publicDER)
	return NodeIdentity{
		NodeID:     "srcnode_" + rawURL.EncodeToString(sum[:]),
		PublicKey:  rawURL.EncodeToString(publicDER),
		PrivateKey: string(pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privateDER})),
	}, nil
}

func ParseEd25519PublicKey(encoded string) (ed25519.PublicKey, []byte, string, error) {
	if len(encoded) < 40 || len(encoded) > 256 {
		return nil, nil, "", errors.New("invalid Ed25519 public key")
	}
	der, err := rawURL.DecodeString(encoded)
	if err != nil {
		return nil, nil, "", errors.New("invalid Ed25519 public key")
	}
	parsed, err := x509.ParsePKIXPublicKey(der)
	if err != nil {
		return nil, nil, "", errors.New("invalid Ed25519 public key")
	}
	key, ok := parsed.(ed25519.PublicKey)
	if !ok || len(key) != ed25519.PublicKeySize {
		return nil, nil, "", errors.New("invalid Ed25519 public key")
	}
	canonical, err := x509.MarshalPKIXPublicKey(key)
	if err != nil || subtle.ConstantTimeCompare(canonical, der) != 1 {
		return nil, nil, "", errors.New("invalid Ed25519 public key")
	}
	return key, canonical, rawURL.EncodeToString(canonical), nil
}

func LoadNodePrivateKey(encoded string) (ed25519.PrivateKey, error) {
	block, _ := pem.Decode([]byte(encoded))
	if block == nil {
		return nil, errors.New("invalid Node private key")
	}
	parsed, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, errors.New("invalid Node private key")
	}
	key, ok := parsed.(ed25519.PrivateKey)
	if !ok {
		return nil, errors.New("invalid Node private key")
	}
	return key, nil
}

func ClientIDFromPublicKey(publicKeyDER []byte) string {
	sum := sha256.Sum256(publicKeyDER)
	return "srcclient_" + rawURL.EncodeToString(sum[:])
}

func UUID() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16]), nil
}

func Base64URL(data []byte) string                 { return rawURL.EncodeToString(data) }
func DecodeBase64URL(value string) ([]byte, error) { return rawURL.DecodeString(value) }
